bulk_insert.go

```diff
@@ -25,79 +25,58 @@ import (
 func BulkInsert(db *gorm.DB, objects []interface{}, chunkSize int, excludeColumns ...string) error {
 	// Split records with specified size not to exceed Database parameter limit
 	for _, objSet := range splitObjects(objects, chunkSize) {
-		if err := insertObjSet(db, objSet, excludeColumns...); err != nil {
+		_, err := insertObjSet(db, objSet, excludeColumns...)
+		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-// BulkInsertWithAssigningIDs executes the query to insert multiple records at once.
-// it will scan the result of `returning id` or `returning *` to [returnedValue] after every insert.
-// it's necessary to set "gorm:insert_option"="returning id" in *gorm.DB
+// BulkInsertWithReturningValues executes the query to insert multiple records at once.
+// This will scan the returned values into `returnedVals`.
 //
-// [returnedValue] slice of primary_key or model, must be a *[]uint(for integer), *[]string(for uuid), *[]struct(for `returning *`)
+// [db] must be set with "gorm:insert_option" to execute RETURNING clause. e.g. db.Set("gorm:insert_option", "RETURNING id")
 //
 // [objects] must be a slice of struct.
 //
+// [returnedVals] must be a point to a slice of struct. Values returned from `RETURNING` clause will be assigned.
+//
 // [chunkSize] is a number of variables embedded in query. To prevent the error which occurs embedding a large number of variables at once
 // and exceeds the limit of prepared statement. Larger size normally leads to better performance, in most cases 2000 to 3000 is reasonable.
 //
 // [excludeColumns] is column names to exclude from insert.
-func BulkInsertWithAssigningIDs(db *gorm.DB, returnedValue interface{}, objects []interface{}, chunkSize int, excludeColumns ...string) error {
-	typ := reflect.TypeOf(returnedValue)
-	if typ.Kind() != reflect.Ptr || typ.Elem().Kind() != reflect.Slice {
-		return errors.New("returningId must be a slice ptr")
+func BulkInsertWithReturningValues(db *gorm.DB, objects []interface{}, returnedVals interface{}, chunkSize int, excludeColumns ...string) error {
```

Comment:
If the interface will only work for returning values and only work for PostgreSQL, would it make sense to convert the `returnedVals` struct fields to table names and set the `insert_option` for the caller?

```go
var returningColumnNames []string

for _, field := range (&gorm.Scope{Value: returnedVals}).Fields() {
    returningColumnNames = append(returningColumnNames, field.DBName)
}

db.Set(
    "gorm:insert_option",
    fmt.Sprintf("RETURNING %s", strings.Join(returningColumnNames, ", "),
)
```

This way we can't get a mismatch between the insert option and the passed struct and if the caller wants to add or remove a column they'll just change the passed struct and it will Just Work:tm:

Reply:
> If the interface will only work for returning values and only work for PostgreSQL

I guess Oracle can also work, though I haven't used and most of the users don't use Oracle for this library😅

Reply:
> By the way, as far as I checked, SQLite also supports.

Cool, didn't know that!